rust/src/res.rs

```diff
@@ -0,0 +1,195 @@
+use ::std::error::Error;
+use ::std::fmt;
+use ::std::fs::File;
+use ::std::io::Seek;
+use ::std::path::Path;
+
+use ::serde::{Deserialize, Serialize};
+use ::serde_json::{json, Map, Value};
+
+use crate::slf::{start_of_entries, SlfHeader, SlfEntryState};
+
+
+// A pack of game resources.
+#[derive(Debug, Default, Deserialize, Serialize)]
+pub struct ResourcePack {
+    // A friendly name of the resource pack for display purposes.
+    pub name: String,
+
+    // The properties of the resource.
+    //
+    // # Properties
+    //
+    //  * with_archive_slf (bool, default = false) - the contents of SLF archives are included
+    pub properties: Map<String,Value>,
+
+    // The resources in this pack.
+    pub resources: Vec<Resource>
+}
+
+
+// A resource in the pack.
+//
+// A resource always maps to raw data, which can be a file or data inside an archive.
+#[derive(Debug, Default, Deserialize, Serialize)]
+pub struct Resource {
+    // The identity of the resource as a relative path.
+    pub path: String,
+
+    // The properties of the resource.
+    //
+    // # Properties
+    //
+    //  * archive_path (string) - the resource path of the archive that contains this resource
+    //  * archive_slf (bool) - true if the resource is a SLF archive
+    //  * archive_slf_num_resources (number) - number of resources inside the SLF archive
+    pub properties: Map<String,Value>
```

Comment:
Could we make the string an enum here, so we don't have arbitrary properties?

Reply:
Or maybe even a `Vec<ResourceProperty>` with

```rust
enum ResourceProperty {
    ArchivePath(PathBuf),
    ArchiveSlf(bool),
...
}
```

Reply:
see https://github.com/ja2-stracciatella/ja2-stracciatella/pull/820#issuecomment-486643636